PeakFinder: OpenCL array overflow spotted by `oclgrind`

```
Invalid write of size 16 at local memory address 0x2000000000030
	Kernel: peakfinder8
	Entity: Global(152,1435,0) Local(0,27,0) Group(152,44,0)
	  store <4 x float> %325, <4 x float> addrspace(3)* %arrayidx256, align 16, !dbg !764
	At line 2197 (column 39) of input.cl:
	  local_peaks[position] = (float4)(som0/sum_int + (float)(gid0), 
	

Invalid read of size 4 at local memory address 0x100000000000c
	Kernel: peakfinder8
	Entity: Global(152,1411,0) Local(0,3,0) Group(152,44,0)
	  %338 = load i32, i32 addrspace(3)* %arrayidx272, align 4, !dbg !793
	At line 2212 (column 34) of input.cl:
	  highidx[write_pos] = local_highidx[tid];
	

Invalid read of size 16 at local memory address 0x2000000000030
	Kernel: peakfinder8
	Entity: Global(152,1411,0) Local(0,3,0) Group(152,44,0)
	  %343 = load <4 x float>, <4 x float> addrspace(3)* %arrayidx276, align 16, !dbg !798
	At line 2213 (column 32) of input.cl:
	  peaks[write_pos] = local_peaks[tid];
	

Invalid write of size 4 at local memory address 0x100000000000c
	Kernel: peakfinder8
	Entity: Global(1835,1425,0) Local(0,17,0) Group(1835,44,0)
	  store i32 %add243, i32 addrspace(3)* %arrayidx245, align 4, !dbg !744
	At line 2196 (column 41) of input.cl:
	  local_highidx[position] = gid0*width + gid1;
	

Invalid write of size 16 at local memory address 0x2000000000030
	Kernel: peakfinder8
	Entity: Global(1835,1425,0) Local(0,17,0) Group(1835,44,0)
	  store <4 x float> %325, <4 x float> addrspace(3)* %arrayidx256, align 16, !dbg !764
	At line 2197 (column 39) of input.cl:
	  local_peaks[position] = (float4)(som0/sum_int + (float)(gid0), 
	

Invalid read of size 4 at local memory address 0x100000000000c
	Kernel: peakfinder8
	Entity: Global(1835,1411,0) Local(0,3,0) Group(1835,44,0)
	  %338 = load i32, i32 addrspace(3)* %arrayidx272, align 4, !dbg !793
	At line 2212 (column 34) of input.cl:
	  highidx[write_pos] = local_highidx[tid];
	

Invalid read of size 16 at local memory address 0x2000000000030
	Kernel: peakfinder8
	Entity: Global(1835,1411,0) Local(0,3,0) Group(1835,44,0)
	  %343 = load <4 x float>, <4 x float> addrspace(3)* %arrayidx276, align 16, !dbg !798
	At line 2213 (column 32) of input.cl:
	  peaks[write_pos] = local_peaks[tid];
	
F
Stderr:
WARNING:pyFAI.ext.splitBBoxCSR:Pixel splitting desactivated !
WARNING:py.warnings:/usr/lib/python3/dist-packages/pytools/py_codegen.py:26: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp

WARNING:py.warnings:/usr/lib/python3/dist-packages/numpy/lib/function_base.py:3405: RuntimeWarning: Invalid value encountered in median
  r = func(a, **kwargs)


======================================================================
FAIL: test_peakfinder8 (pyFAI.opencl.test.test_peak_finder.TestOclPeakFinder)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/kieffer/workspace/pyFAI/build/lib.linux-x86_64-3.7/pyFAI/opencl/test/test_peak_finder.py", line 186, in test_peakfinder8
    self.assertLess(numpy.median(dm.min(axis=1)), 1, "Most peaks are found")
AssertionError: nan not less than 1 : Most peaks are found

Stderr:
WARNING:pyFAI.ext.splitBBoxCSR:Pixel splitting desactivated !
WARNING:py.warnings:/usr/lib/python3/dist-packages/pytools/py_codegen.py:26: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp

WARNING:py.warnings:/usr/lib/python3/dist-packages/numpy/lib/function_base.py:3405: RuntimeWarning: Invalid value encountered in median
  r = func(a, **kwargs)
```

--- a/pyFAI/ext/splitBBoxLUT.pyx
+++ b/pyFAI/ext/splitBBoxLUT.pyx
@@ -37,7 +37,7 @@ reverse implementation based on a sparse matrix multiplication
 
 __author__ = "Jerome Kieffer"
 __contact__ = "Jerome.kieffer@esrf.fr"
-__date__ = "19/11/2021"
+__date__ = "22/11/2021"
 __status__ = "stable"
 __license__ = "MIT"
 
@@ -276,7 +276,7 @@ class HistoBBox1d(LutIntegrator):
             position_t[::1] cpos0_inf = self.cpos0_inf
             position_t[::1] cpos1_min, cpos1_max
             mask_t[::1] cmask
-            SparseBuilder builder = SparseBuilder(bins, block_size=32, heap_size=bins*32)
+            SparseBuilder builder = SparseBuilder(bins, block_size=32, heap_size=size)
 
         size = self.size
         if self.check_mask:
@@ -348,7 +348,7 @@ class HistoBBox1d(LutIntegrator):
             bint check_mask, check_pos1
             position_t[::1] cpos0 = self.cpos0, cpos1_min, cpos1_max,
             mask_t[::1] cmask
-            SparseBuilder builder = SparseBuilder(bins, block_size=32, heap_size=bins*32)
+            SparseBuilder builder = SparseBuilder(bins, block_size=32, heap_size=size)
             
         size = self.size
         if self.check_mask:
@@ -594,7 +594,7 @@ class HistoBBox2d(LutIntegrator):
             lut_t[:, :, ::1] lut
             mask_t[:] cmask
             acc_t inv_area, delta_down, delta_up, delta_right, delta_left
-            SparseBuilder builder = SparseBuilder(bins0*bins1, block_size=6, heap_size=bins0*bins1)
+            SparseBuilder builder = SparseBuilder(bins0*bins1, block_size=8, heap_size=size)
 
         if self.check_mask:
             cmask = self.cmask

--- a/pyFAI/ext/splitBBoxLUT.pyx
+++ b/pyFAI/ext/splitBBoxLUT.pyx
@@ -270,7 +270,7 @@ class HistoBBox1d(LutIntegrator):
         cdef:
             position_t delta = self.delta, pos0_min = self.pos0_min, pos1_min, pos1_max, min0, max0, fbin0_min, fbin0_max
             acc_t delta_left, delta_right, inv_area
-            int k, idx, bin0_min, bin0_max, bins = self.bins, lut_size, i, size
+            int k, idx, bin0_min, bin0_max, bins = self.bins, lut_size, i, size = self.size
             bint check_mask, check_pos1
             position_t[::1] cpos0_sup = self.cpos0_sup
             position_t[::1] cpos0_inf = self.cpos0_inf
@@ -278,7 +278,7 @@ class HistoBBox1d(LutIntegrator):
             mask_t[::1] cmask
             SparseBuilder builder = SparseBuilder(bins, block_size=32, heap_size=size)
 
-        size = self.size
+        
         if self.check_mask:
             cmask = self.cmask
             check_mask = True
@@ -344,13 +344,12 @@ class HistoBBox1d(LutIntegrator):
 
         cdef:
             position_t delta = self.delta, pos0_min = self.pos0_min, pos1_min, pos1_max, fbin0, pos0
-            int32_t k, idx, bin0, bins = self.bins, size, nnz
+            int32_t k, idx, bin0, bins = self.bins, nnz, size = self.size
             bint check_mask, check_pos1
             position_t[::1] cpos0 = self.cpos0, cpos1_min, cpos1_max,
             mask_t[::1] cmask
             SparseBuilder builder = SparseBuilder(bins, block_size=32, heap_size=size)
-            
-        size = self.size
+
         if self.check_mask:
             cmask = self.cmask
             check_mask = True

--- a/pyFAI/ext/splitPixelFullCSR.pyx
+++ b/pyFAI/ext/splitPixelFullCSR.pyx
@@ -1,8 +1,8 @@
 # -*- coding: utf-8 -*-
 #cython: embedsignature=True, language_level=3, binding=True
-##cython: boundscheck=False, wraparound=False, cdivision=True, initializedcheck=False,
+#cython: boundscheck=False, wraparound=False, cdivision=True, initializedcheck=False,
 ## This is for developping
-# cython: profile=True, warn.undeclared=True, warn.unused=True, warn.unused_result=False, warn.unused_arg=True
+## cython: profile=True, warn.undeclared=True, warn.unused=True, warn.unused_result=False, warn.unused_arg=True
 #
 #    Project: Fast Azimuthal Integration
 #             https://github.com/silx-kit/pyFAI
@@ -35,7 +35,7 @@ Sparse matrix represented using the CompressedSparseRow.
 
 __author__ = "Jerome Kieffer"
 __contact__ = "Jerome.kieffer@esrf.fr"
-__date__ = "19/11/2021"
+__date__ = "22/11/2021"
 __status__ = "stable"
 __license__ = "MIT"
 
@@ -249,7 +249,7 @@ class FullSplitCSR_1d(CsrIntegrator):
             Function AB, BC, CD, DA
             Py_ssize_t bins=self.bins, i = 0, idx = 0, bin = 0, bin0 = 0, bin0_max = 0, bin0_min = 0, k = 0, size = 0, pos=0
             bint check_pos1, check_mask = False
-            SparseBuilder builder = SparseBuilder(bins, block_size=32, heap_size=bins*32)
+            SparseBuilder builder = SparseBuilder(bins, block_size=32, heap_size=size)
 
         if self.pos0_range is not None:
             self.pos0_min, self.pos0_maxin = self.pos0_range
@@ -477,7 +477,7 @@ class FullSplitCSR_2d(CsrIntegrator):
             buffer_t[::1] linbuffer = numpy.empty(256, dtype=buffer_d)
             buffer_t[:, ::1] buffer = numpy.asarray(linbuffer[:(bw0+1)*(bw1+1)]).reshape((bw0+1,bw1+1))
             double foffset0, foffset1, sum_area, loc_area
-            SparseBuilder builder = SparseBuilder(bins1*bins0, block_size=8, heap_size=bins1*bins0)
+            SparseBuilder builder = SparseBuilder(bins1*bins0, block_size=8, heap_size=size)
             
         if check_mask:
             cmask = self.cmask

--- a/pyFAI/opencl/azim_csr.py
+++ b/pyFAI/opencl/azim_csr.py
@@ -28,14 +28,13 @@
 
 __authors__ = ["Jérôme Kieffer", "Giannis Ashiotis"]
 __license__ = "MIT"
-__date__ = "28/10/2021"
+__date__ = "24/11/2021"
 __copyright__ = "2014-2021, ESRF, Grenoble"
 __contact__ = "jerome.kieffer@esrf.fr"
 
 import logging
 from collections import OrderedDict
 import numpy
-import math
 from . import pyopencl
 from ..utils import calc_checksum
 if pyopencl:

--- a/pyFAI/opencl/peak_finder.py
+++ b/pyFAI/opencl/peak_finder.py
@@ -29,7 +29,7 @@
 
 __authors__ = ["Jérôme Kieffer"]
 __license__ = "MIT"
-__date__ = "22/11/2021"
+__date__ = "23/11/2021"
 __copyright__ = "2014-2021, ESRF, Grenoble"
 __contact__ = "jerome.kieffer@esrf.fr"
 
@@ -745,9 +745,9 @@ class OCL_PeakFinder(OCL_CSR_Integrator):
         wdim_data = (data.shape[0] + wg0 - 1) & ~(wg0 - 1), (data.shape[1] + wg1 - 1) & ~(wg1 - 1)
         # allocate local memory: we store 4 bytes but at most 1 pixel out of 4 can be a peak
         
-        buffer_size = int(math.ceil(wg * 4 / (patch_size * patch_size)))
-        print(wg, wdim_data, buffer_size)
-        kw_proj["local_highidx"] = pyopencl.LocalMemory(4 * buffer_size)
+        buffer_size = int(math.ceil(wg * 4 / (patch_size)))
+        print(wg, wg0, wg1, wdim_data, buffer_size)
+        kw_proj["local_highidx"] = pyopencl.LocalMemory(1 * buffer_size)
         kw_proj["local_peaks"] = pyopencl.LocalMemory(4 * buffer_size)
         kw_proj["local_buffer"] = pyopencl.LocalMemory(8 * (wg0 + 2 * hw) * (wg1 + 2 * hw))
         kw_proj["half_patch"] = numpy.int32(hw)

--- a/pyFAI/opencl/peak_finder.py
+++ b/pyFAI/opencl/peak_finder.py
@@ -29,7 +29,7 @@
 
 __authors__ = ["Jérôme Kieffer"]
 __license__ = "MIT"
-__date__ = "18/11/2021"
+__date__ = "22/11/2021"
 __copyright__ = "2014-2021, ESRF, Grenoble"
 __contact__ = "jerome.kieffer@esrf.fr"
 
@@ -182,6 +182,8 @@ class OCL_PeakFinder(OCL_CSR_Integrator):
                                                           ("width", numpy.int32(2)),
                                                           ("half_patch", numpy.int32(1)),
                                                           ("connected", numpy.int32(2)),
+                                                          # ("max_pkg", numpy.int32(2)),
+                                                          # ("max_pkl", numpy.int32(2)),
                                                           ("counter", self.cl_mem["counter"]),
                                                           ("peak_position", self.cl_mem["peak_position"]),
                                                           ("peak_descriptor", self.cl_mem["peak_descriptor"]),
@@ -742,8 +744,10 @@ class OCL_PeakFinder(OCL_CSR_Integrator):
 
         wdim_data = (data.shape[0] + wg0 - 1) & ~(wg0 - 1), (data.shape[1] + wg1 - 1) & ~(wg1 - 1)
         # allocate local memory: we store 4 bytes but at most 1 pixel out of 4 can be a peak
+        
         buffer_size = int(math.ceil(wg * 4 / (patch_size * patch_size)))
-        kw_proj["local_highidx"] = pyopencl.LocalMemory(buffer_size)
+        print(wg, wdim_data, buffer_size)
+        kw_proj["local_highidx"] = pyopencl.LocalMemory(4 * buffer_size)
         kw_proj["local_peaks"] = pyopencl.LocalMemory(4 * buffer_size)
         kw_proj["local_buffer"] = pyopencl.LocalMemory(8 * (wg0 + 2 * hw) * (wg1 + 2 * hw))
         kw_proj["half_patch"] = numpy.int32(hw)

--- a/pyFAI/opencl/peak_finder.py
+++ b/pyFAI/opencl/peak_finder.py
@@ -29,7 +29,7 @@
 
 __authors__ = ["Jérôme Kieffer"]
 __license__ = "MIT"
-__date__ = "23/11/2021"
+__date__ = "24/11/2021"
 __copyright__ = "2014-2021, ESRF, Grenoble"
 __contact__ = "jerome.kieffer@esrf.fr"
 
@@ -728,6 +728,7 @@ class OCL_PeakFinder(OCL_CSR_Integrator):
             kw_proj["radius_min"] = numpy.float32(0.0)
             kw_proj["radius_max"] = numpy.float32(numpy.finfo(numpy.float32).max)
 
+        print(self.workgroup_size["peakfinder8"])
         wg = max(self.workgroup_size["peakfinder8"])
         if wg >= 32:
             wg1 = 32
@@ -745,7 +746,7 @@ class OCL_PeakFinder(OCL_CSR_Integrator):
         wdim_data = (data.shape[0] + wg0 - 1) & ~(wg0 - 1), (data.shape[1] + wg1 - 1) & ~(wg1 - 1)
         # allocate local memory: we store 4 bytes but at most 1 pixel out of 4 can be a peak
         
-        buffer_size = int(math.ceil(wg * 4 / (patch_size)))
+        buffer_size = int(math.ceil(wg * 4 / ((1+hw)*min(wg0, 1+hw))))
         print(wg, wg0, wg1, wdim_data, buffer_size)
         kw_proj["local_highidx"] = pyopencl.LocalMemory(1 * buffer_size)
         kw_proj["local_peaks"] = pyopencl.LocalMemory(4 * buffer_size)

--- a/pyFAI/opencl/peak_finder.py
+++ b/pyFAI/opencl/peak_finder.py
@@ -695,7 +695,6 @@ class OCL_PeakFinder(OCL_CSR_Integrator):
             self.cl_kernel_args["corrections4"]["do_mask"] = numpy.int8(1)
         wg = max(self.workgroup_size["corrections4"])
         wdim_data = (self.size + wg - 1) & ~(wg - 1),
-#         print(kw_corr)
         ev = self.kernels.corrections4(self.queue, wdim_data, (wg,), *list(kw_corr.values()))
         events.append(EventDescription("corrections", ev))
 
@@ -728,9 +727,8 @@ class OCL_PeakFinder(OCL_CSR_Integrator):
             kw_proj["radius_min"] = numpy.float32(0.0)
             kw_proj["radius_max"] = numpy.float32(numpy.finfo(numpy.float32).max)
 
-        print(self.workgroup_size["peakfinder8"])
         wg = max(self.workgroup_size["peakfinder8"])
-        if wg >= 32:
+        if wg > 32:
             wg1 = 32
             wg0 = wg // 32
         else:
@@ -747,7 +745,6 @@ class OCL_PeakFinder(OCL_CSR_Integrator):
         # allocate local memory: we store 4 bytes but at most 1 pixel out of 4 can be a peak
         
         buffer_size = int(math.ceil(wg * 4 / ((1+hw)*min(wg0, 1+hw))))
-        print(wg, wg0, wg1, wdim_data, buffer_size)
         kw_proj["local_highidx"] = pyopencl.LocalMemory(1 * buffer_size)
         kw_proj["local_peaks"] = pyopencl.LocalMemory(4 * buffer_size)
         kw_proj["local_buffer"] = pyopencl.LocalMemory(8 * (wg0 + 2 * hw) * (wg1 + 2 * hw))

--- a/pyFAI/resources/openCL/peakfinder8.cl
+++ b/pyFAI/resources/openCL/peakfinder8.cl
@@ -127,12 +127,14 @@ kernel void peakfinder8(  const global  float4 *preproc4, // pixel wise array of
                           const         int     width,     // width of the image
                           const         int     half_patch,// half of the size of the patch, 1 or 2 for 3x3 or 5x5 patch
                           const         int     connected, // keep only enough pixels are above threshold in patch
+//                          const         int    max_pkg,    // size of the peak array in global memory
+//                          const         int    max_pkl,    // size of the peak array in local memory
                                 global  int    *counter,   // Counter of the number of peaks found
                                 global  int    *highidx,   // indexes of the pixels of high intensity
                                 global  float4 *peaks,     // Contains center0, center1, integrated, sigma 
                                 local   int    *local_highidx,   // size: wg0*wg1
                                 local   float4 *local_peaks, // size: wg0*wg1
-                                local   float2 *buffer){         // size: (wg0+2*half_patch)*(wg1+2*half_patch) 
+                                local   float2 *buffer){   // size: (wg0+2*half_patch)*(wg1+2*half_patch) 
     int tid0 = get_local_id(0);
     int tid1 = get_local_id(1);
     int gid0 = get_global_id(0);

--- a/pyFAI/resources/openCL/peakfinder8.cl
+++ b/pyFAI/resources/openCL/peakfinder8.cl
@@ -113,15 +113,50 @@ inline float2 get_shared2(local float2* buffer, int dim0, int dim1, int half_pat
     return buffer[pf8_calc_buffer2(dim0, dim1, half_patch)];
 }
 
-
-kernel void peakfinder8(  const global  float4 *preproc4, // pixel wise array of (signal, variance, norm, cnt) 
-                          const global  float  *radius2d, // contains the distance to the center for each pixel
+/*peakfinder8: kernel that pick peaks in an image after the sigma-clipping
+ * 
+ * In this kernel, each thread works on one pixel and uses data from the neighborhood of radius `half_patch` size
+ * The sigma-clipping step preceeding provides the mean signal and the associated uncertainty.
+ * This uncertainty sees some noise added quadratically.   
+ * 
+ * A pixel is kepts if: 
+ *  - if it a local maximum, 
+ *  - its intensity is greater than `mean` plus `cutoff`*`std` (including noise)
+ *  - more than `connected` pixels are also intense in the direct vinicy defined by a radius of `half_patch` size.
+ *  
+ * For each peak is calculated:
+ *  - the local sum of intensity over the vinicy, only accounting intense pixels
+ *  - the associated uncertainty (variance propagation).
+ *  - the center of mass of the peak 
+ *  
+ * The workgroup size should be as square as possible since the kernel first load all information and uses an explicit stencil pattern   
+ *
+ * parameters:
+ *  - preproc4: pixel wise array of (signal, variance, norm, cnt)
+ *  - radius2d: Contains the distance to the beamcenter for each pixel
+ *  - radius1d: Radial bin postion
+ *  - average1d: average intensity for pixels at the given distance from beam-center
+ *  - std1d: uncertainty associated to average1d
+ *  - radius_min: inner-most radius considered (not in bin, but actual radius)
+ *  - radius_min: outer-most radius considered (not in bin, but actual radius)
+ *  - cutoff: Intense pixels are the onve with I > avg + cutoff*√(std²+noise²)
+ *  - noise: noise level of the measurement, added quadratically to the uncertainty
+ *  - heigth: heigth of the image
+ *  - width: width of the image
+ *  - half_patch: radius or half of the size of the patch to consider for the neighborhood definition. Usually 1 or 2 to have a 3x3 or 5x5 patch.
+ *  - connected: keep peak if so many intense pixels are found in the patch
+ *  - counter: Number of peaks found
+ *  - highidx: index of the most intense pixel of each peak
+ *  - peaks: array of struct containing the centroid0, centroid1, sum of intensity and associated uncertainty calculated over all intensi pixels in the patch
+ */
+kernel void peakfinder8(  const global  float4 *preproc4, // Pixel wise array of (signal, variance, norm, cnt) 
+                          const global  float  *radius2d, // Contains the distance to the center for each pixel
                           const global  float  *radius1d, // Radial bin postion 
                           const global  float  *average1d,// average intensity in the bin
                           const global  float  *std1d,    // associated deviation
                           const         float   radius_min,// minimum radius
                           const         float   radius_max,// maximum radius 
-                          const         float   cutoff,    // pick pixel with I>avg+min(cutoff*std, noise)
+                          const         float   cutoff,    // pick pixel with I>avg+cutoff*√(std²+noise²)
                           const         float   noise,     // Noise level of the measurement
                           const         int     heigth,    // heigth of the image
                           const         int     width,     // width of the image

